### DISCUSSION TITLE
Astro Action Generators

### DISCUSSION THREAD

**Discussion by dnbkr** [withastro/roadmap#1059] [Proposal]
# Summary

Allow for a new type of Astro Action - one that is implemented with an asynchronous generator function and delivered to the client as an EventSource (SSE)

# Background & Motivation

Currently in the astro app I am working on we prefer to use Astro Actions for our asynchronous frontend-to-server communication, because of its type-safety and lack of boilerplate. However we do stream some data to the client using server-sent events and currently we use API routes for this. It seems to me that it would fit quite nicely with the Actions paradigm where the developer only really needs to think about writing a function and then can call that function from the frontend and the intermediate communication is abstracted/hidden away.

# Goals

To allow the server to stream data to the frontend with the simplicity of Astro Actions.

# Example

```typescript
const action = defineAction({
  input: z.object({ id: z.string() }),
  handler: async function*({ id }) {
    const items = await loadItemsFromDB(id);
    for (const item of items) {
      await processItem(item);
      yield `${item.name} Processed`
    }
  }
})
```

```typescript
const stream = actions.processItems({ id: 'id' });

for await (let progress of stream) {
  console.log(progress)
}
```

# Links

- [MDN: EventSource](https://developer.mozilla.org/en-US/docs/Web/API/EventSource)
- [MDN: Async Generator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/AsyncGenerator)

**Comment by ematipico** [comment]
Generators are slow unfortunately and we can't do much to make them faster. We internally used to use generators, and we removed them to speed up code for users.

**Reply by dnbkr** [reply]
in the use case I am imagining, would this not be a long running task (hence the need to stream a response to the client) and therefore the overhead of a generator is less of a concern?

EDIT: for clarity, I am not suggesting _all_ actions use generators, only if you wanted to stream a response back with an eventsource

**Reply by mszajna** [reply]
I'm very curious about performance issues you've encounter with generators. All I could find was the significant performance hit incurred when transforming synchronous collections into async generators, as the latter require event loop for each item. Surely an issue for CLI applications, but not sure if that much of a problem when you've got a wire in the middle anyway.

Would you be more open to considering a non-generator-based API to achieve streaming with actions? Perhaps such that could be adapted to generators if the user choses to opt into?

**Reply by delucis** [reply]
Yeah, the generator performance issue we ran into was in hot path rendering code. I would guess in a process that is already slow like this suggested use case it wouldn’t matter as much.

(That said, I’m not 100% sure actions are the right fit for trying to build an API for long-lived data sources like server-sent events.)
